File: fixtures/uniffi-fixture-time/tests/bindings/test_chronological.py
from chronological import *
from datetime import datetime, timedelta, timezone, MAXYEAR
import sys

# Test passing timestamp and duration while returning timestamp
assert add(datetime.fromtimestamp(100.000001), timedelta(seconds=1, microseconds=1)).timestamp() == 101.000002

# Test passing timestamp while returning duration
assert diff(datetime.fromtimestamp(101.000002), datetime.fromtimestamp(100.000001)) == timedelta(seconds=1, microseconds=1)

# Test exceptions are propagated
try:
    diff(datetime.fromtimestamp(100), datetime.fromtimestamp(101))
    assert(not("Should have thrown a TimeDiffError exception!"))
except ChronologicalError.TimeDiffError:
    # It's okay!
    pass

# Test unix epoch lower bound
try:
    diff(datetime.fromtimestamp(-1), datetime.fromtimestamp(101))
    assert(not("Should have thrown a ValueError exception!"))
except ValueError:
    # It's okay!
    pass

# Test near max timestamp bound, no microseconds due to python floating point precision issues
assert add(datetime(MAXYEAR, 12, 31, 23, 59, 59, 0), timedelta(seconds=0)) == datetime(MAXYEAR, 12, 31, 23, 59, 59, 0)

# Test overflow at max timestamp bound
try:
    add(datetime(MAXYEAR, 12, 31, 23, 59, 59, 0), timedelta(seconds=1))
    assert(not("Should have thrown a ValueError exception!"))
except ValueError:
    # It's okay!
    pass

# Test that rust timestamps behave like kotlin timestamps
pythonBefore = datetime.now()
rustNow = now()
pythonAfter = datetime.now()

assert pythonBefore <= rustNow <= pythonAfter

# Test that uniffi returns naive datetime
assert now().tzinfo is None

File: fixtures/uniffi-fixture-time/tests/bindings/test_chronological.swift
import Foundation
import chronological

// Test passing timestamp and duration while returning timestamp
assert(try! add(a: Date.init(timeIntervalSince1970: 100.01), b: 1.01) == Date.init(timeIntervalSince1970: 101.02), "add duration")

// Test passing timestamp while returning duration (note precision error)
assert(try! diff(a: Date.init(timeIntervalSince1970: 101.03), b: Date.init(timeIntervalSince1970: 100.01)) == 1.019999981, "diff dates")

// Test pre-epoch timestamps
let iso8601 = ISO8601DateFormatter()
iso8601.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
assert(try! add(a: iso8601.date(from: "1955-11-05T00:06:00.283Z")!, b: 1.001) == iso8601.date(from: "1955-11-05T00:06:01.284Z")!, "pre-epoch add")

// Test exceptions are propagated
do {
    let _ = try diff(a: Date.init(timeIntervalSince1970: 100), b: Date.init(timeIntervalSince1970: 101))
    fatalError("Should have thrown a TimeDiffError exception!")
} catch ChronologicalError.TimeDiffError {
    // It's okay!
}

// Test that rust timestamps behave like swift timestamps.
// Sleep briefly between each call: Date uses Double (CFAbsoluteTime) which has
// ~100–200 ns precision at current epoch, so without a gap the ordering can
// flip intermittently due to floating-point rounding in the FFI conversion.
let swiftBefore = Date.init()
Thread.sleep(forTimeInterval: 0.01)
let rustNow = now()
Thread.sleep(forTimeInterval: 0.01)
let swiftAfter = Date.init()

assert(swiftBefore < rustNow)
assert(swiftAfter > rustNow)